File: lab2/ClassArray.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace lab2
{
    class ClassArray<T> where T : IAnimals
    {
        //4
        private Dictionary<int, T> cells;
        private int maxCount;
        //

        private T defaultValue;
        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            //4
            cells = new Dictionary<int, T>();
            maxCount = size;
            //
        }

        public static int operator +(ClassArray<T> c, T shark)
        {  //4
            if (c.cells.Count == c.maxCount)
            {
                return -1;
            }
            //
            for (int i = 0; i < c.cells.Count; i++)
            {
                if (c.ChekFreeCell(i))
                {
                    //4
                    c.cells.Add(i, shark);
                    //
                    return i;
                }
            }
            //4
            c.cells.Add(c.cells.Count, shark);
            return c.cells.Count - 1;
            //
        }

        public static T operator -(ClassArray<T> c, int index)
        {
            //4
            if (c.cells.ContainsKey(index))
            {
                T shark = c.cells[index];
                c.cells.Remove(index);
                return shark;
            }
            //
            return c.defaultValue;
        }

        private bool ChekFreeCell(int index)
        {
            //4
            return !cells.ContainsKey(index);
            //
        }
        //4 
        public T this[int ind]
        {
            get
            {
                if (cells.ContainsKey(ind))
                {
                    return cells[ind];
                }
                return defaultValue;
            }
        }
        //
        public T getObject(int ind)
        {
            if (ind > -1 && ind < cells.Count)
            {
                return cells[ind];
            }
            return defaultValue;
        }
    }
}

File: src/ClassArray.java
class ClassArray<T> {
	private T IAnimal;
	private IAnimal[] cells;

	private IAnimal defaultValue;

	public ClassArray(int size, IAnimal defVal) {
		defaultValue = defVal;
		cells = new IAnimal[size];
		for (int i = 0; i < cells.length; i++) {
			cells[i] = defaultValue;
		}
	}

	public static int Plus(ClassArray<IAnimal> c, IAnimal shark) {
		for (int i = 0; i < c.cells.length; i++) {
			if (c.ChekFreeCell(i)) {
				c.cells[i] = shark;
				return i;
			}
		}
		return -1;
	}

	public static IAnimal Minus(ClassArray<IAnimal> c, int index) {
		if (!c.ChekFreeCell(index)) {
			IAnimal shark = c.cells[index];
			c.cells[index] = c.defaultValue;
			return shark;
		}
		return c.defaultValue;
	}

	private boolean ChekFreeCell(int index) {
		if (index < 0 || index > cells.length) {
			return false;
		}
		if (cells[index] == null) {
			return true;
		}
		if (cells[index].equals(defaultValue)) {
			return true;
		}
		return false;
	}

	public IAnimal getObject(int ind) {
		if (ind > -1 && ind < cells.length) {
			return cells[ind];
		}
		return defaultValue;
	}
}